quinn-proto/src/connection/assembler.rs

```diff
@@ -193,6 +228,48 @@ impl Chunk {
 struct Buffer {
     offset: u64,
     bytes: Bytes,
+    allocation_size: usize,
+    defragmented: bool,
+}
+
+impl Buffer {
+    /// Constructs a new fragmented Buffer
+    fn new(offset: u64, bytes: Bytes, allocation_size: usize) -> Self {
+        Self {
+            offset,
+            bytes,
+            allocation_size,
+            defragmented: false,
+        }
+    }
+
+    /// Constructs a new defragmented Buffer
+    fn new_defragmented(offset: u64, bytes: Bytes) -> Self {
+        let allocation_size = bytes.len();
+        Self {
+            offset,
+            bytes,
+            allocation_size,
+            defragmented: true,
+        }
+    }
+
+    /// Discards data before `offset` and flags `self` as defragmented if it has good utilization
+    fn prepare_defragment(&mut self, offset: u64) {
```

Comment:
Nit: I would potentially call this `try_mark_defragmented`, since it changes the state already without anything after it. `prepare` sounds like there would be a follow-up call to get it into the state.

Reply:
Done.